src/app/components/Account/AccountLinkWithAddressSwitch.tsx

```diff
@@ -0,0 +1,39 @@
+import { FunctionComponent, useEffect, useState } from 'react'
+import { AddressSwitchOption } from '../AddressSwitch'
+import { AccountLink, AccountLinkProps } from './AccountLink'
+import { getOasisAddress, isValidEthAddress, isValidOasisAddress } from '../../utils/helpers'
+
+interface WithAddressSwitchProps {
+  addressSwitchOption: AddressSwitchOption
+}
+
+interface WrappedComponentProps extends AccountLinkProps, WithAddressSwitchProps {}
+
+const withAddressSwitch =
+  (Component: FunctionComponent<AccountLinkProps>): FunctionComponent<WrappedComponentProps> =>
+  (props: WrappedComponentProps) => {
+    const { addressSwitchOption, address, ...restProps } = props
+
+    const [convertedAddress, setConvertedAddress] = useState<string>('')
+
+    useEffect(() => {
+      const resolveAddresses = async () => {
+        if (addressSwitchOption === AddressSwitchOption.Oasis && isValidOasisAddress(address)) {
+          setConvertedAddress(address)
+        } else if (addressSwitchOption === AddressSwitchOption.ETH && isValidEthAddress(address)) {
+          setConvertedAddress(address)
+        } else if (addressSwitchOption === AddressSwitchOption.Oasis && isValidEthAddress(address)) {
+          const oasisAddress = await getOasisAddress(address)
+          setConvertedAddress(oasisAddress)
+        } else {
+          setConvertedAddress(address)
+        }
+      }
+
+      resolveAddresses()
+    }, [address, addressSwitchOption])
+
+    return <Component {...restProps} address={convertedAddress} />
+  }
```

Comment:
Should this be simplified down to accepting two types of addresses, like `ethAddress` and `oasisAddress` and returning address based on the `addressSwitchOption`. 

Or is there a case where only ETH address is available, and it needs conversion to Oasis address. Because that is the only benefit of this code, over the previous mentioned "improvement".

Reply:
Simplified those places: https://github.com/oasisprotocol/explorer/pull/1365

Reply:
Amazing 👍 This simplifies the code a lot. Will wait till #1365 is merged first.

Reply:
Did remove manually conversion, so now it only switches the address and default to Oasis type address for default, if ETH type address is not available.